Add `Part-of:` trailer to patches

demize opened the pull request:
I went with `Part-of:` because it feels the most broadly correct (across forges).

To achieve this, I refactored the loop where it checks signoffs to also optionally (but by default) add `Part-of:` trailers.

I think this also fixes a bug where the patches weren't looped through correctly, so only the first commit was checked for signoffs.

This also fixes a minor documentation issue, where the git config key for `pram.repo` was listed as `pram.repository`.

(Truthfully, this was initially just to fix that documentation issue, but then I saw an open issue I could tackle while I was at it :)

demize committed:
Add `Part-of:` trailer to patches

I went with `Part-of:` because it feels the most broadly
correct (across forges).

To achieve this, I refactored the loop where it checks
signoffs to also optionally (but by default) add `Part-of:`
trailers.

I think this also fixes a bug where the patches weren't
looped through correctly, so only the first commit was
checked for signoffs.

This also fixes a minor documentation issue, where the git
config key for `pram.repo` was listed as `pram.repository`.

Closes: #7
Signed-off-by: Vivian Heisz (demize) <demize@unstable.systems>

mgorny commented on pram:
```diff
@@ -242,11 +256,17 @@ main() {
 
 		local patches=( "${tempdir}"/[0-9]* )
-		if [[ ${signoff} ]]; then
+		if [[ ${signoff} ]] || [[ ${partof} ]]; then
```
```suggestion
		if [[ ${signoff} || ${partof} ]]; then
```

mgorny reviewed (changes requested):
Thanks for doing this. However, could you add a test case too?

CC @thesamesam, he definitely wanted something like this.

mgorny commented:
I think we should enable it by default, or maybe even unconditionally (let's wait to see what others think).

thesamesam reviewed (approved):
Thank you for doing this! It looks good to me once mgorny's suggestion is included and we also have some tests ideally.

As for default: IMO default on, makes sense to have it conditional in case you pulled it from some temporary fork or PR or similar (I sometimes use `pram` for that).

demize committed:
Add tests for `--part-of` and ensure it works locally

This also adds a `--part-of-pr` option to override the
linked PR, specifically for use with local files. I'm not
-thrilled- with this solution, but I tried a few different
ways and this ended up feeling the best: it lets users set
their default to not add `Part-of:` but still use `-p` to
enable it without manually specifying a PR, it allows you
to add the trailer with local patches, and it lets me write
a test for it.

Signed-off-by: Vivian Heisz (demize) <demize@unstable.systems>

demize committed:
Add test for partial signoff rejection

Signed-off-by: Vivian Heisz (demize) <demize@unstable.systems>

demize committed:
Add `-P` option to most tests

With the addition of the `Part-of` trailer by default,
most tests were broken. The easiest fix is to disable
the trailer for them--I added a test for the trailer,
so we should know it's working fine...

Signed-off-by: Vivian Heisz (demize) <demize@unstable.systems>

demize commented:
Turns out adding tests made me think a lot more about how this was implemented--tests only really work (from what I can see) with local patches, so they necessitated a way to override the linked pull request. It also got me to fix the rest of the tests, which I accidentally broke...

I think I'm happy with what I've done here, but it's pretty late here, so I'll gladly take feedback on the surprising amount of changes I just pushed!

mgorny commented on pram:
```diff
@@ -149,6 +156,24 @@ main() {
+				user_pr=${2}
+				shift
+				;;
+			--part-of-pr=*)
```
In general, I think it's a good idea. But I think it should also override what we put as `Closes:` implicitly, i.e. we are effectively having a series of `Part-of` trailers, followed by a `Closes` trailer that should always be referencing the same bug.

As such, we probably need a better name too.

mgorny commented on pram:
```diff
@@ -149,6 +156,24 @@ main() {
+				[[ -z ${user_pr} ]] || die "${0}: cannot specify multiple ${1%%=*}"
+				user_pr=${1#*=}
+				;;
+			-P|--no-part-of)
```
Similarly, I wonder if this option shouldn't also disabling adding the implicit `Closes`.

mgorny commented on pram:
```diff
@@ -203,31 +233,36 @@ main() {
 
-	local pr to_close
+	local pr to_close linked_pr
 	for pr in "${prs[@]}"; do
```
Note that we technically permit passing multiple pull requests — should the `Part-of` override force the same PR for all of them? Perhaps we should error out when it's passed with multiple PRs, and revisit if somebody actually complains.

demize replied:
I think an interesting alternate approach here (that might make more sense overall) would be to target local patch files more specifically. Drop the `--part-of-pr` option again, make the argument part of `-p`, and build an array out of the `-p` arguments to use with local patch files. It would still let you override the implicit PR for other options, and `-P` would still disable entirely. This approach would mean that if you disable the trailer in the git config, you can no longer use the implicit PR at all, but I'm not sure that's a bad thing. 

This would ultimately let you merge multiple PRs with `pram -p 123 a.patch -p 455 b.patch`, which I think is a pretty natural way to do it. If you're specifying PR numbers you'd need to do it for *every* link but I think that's probably fine?

This would also work nicely with your other comment about overriding `to_close`. Instead of `--part-of`, call this `--pull-request`, and make it explicit that it overrides what `Closes` trailer it adds too.

Shouldn't be much of an issue for me to rework it this way, if you agree with the approach.

mgorny replied:
Mixed feelings here. It's a major change of behavior and context-sensitive options tend to be surprising. If we do that, we should probably make other options context-sensitive too, and that's a potentially breaking change.

On the other hand, I honestly doubt many people apply multiple patches at once. Perhaps the simplest solution would be to remove that entirely, and allow only a single patch at a time.

thesamesam replied:
I had no idea pram could do that and I expect others don't either. Let's drop it and see if anyone complains. It's IMO too much to add all of this unless we know there's a need for it.

demize replied:
Would that be dropping support for multiple local patch files, or for multiple patches entirely? I should be able to do either, add it on top of this PR.

mgorny replied:
Let's drop it entirely. It would be probably cleaner to start from `master` and probably do one thing at a time, since there would be too much back-and-forth in the commit history. Please let me know if you want to work on it or if I should do it myself first.

demize replied:
I'll work on it this afternoon, have a new set of changes sent in today. Agreed that it'll be cleaner to start fresh, so it might be a second PR. I'll close this one if I have to open a second.

mgorny replied:
Thanks. Ideally, do a single change (removal of multi-PR support) first and let's get that merged before you go further. I really feel bad when you have to rebase all your work because of changes requested.

demize replied:
Done, opened #10. Since that's open now I'll close this, and once it's landed I'll open a third PR with the rest of the changes (organized a little better, this time :)

mgorny commented on test/00basic.sh:
```diff
@@ -43,7 +43,7 @@ cat > trivial.patch <<-EOF
 EOF
 
-bash "${INITDIR}"/../pram --no-gitconfig -e true -G -I -S ./trivial.patch
+bash "${INITDIR}"/../pram --no-gitconfig -e true -G -I -S -P ./trivial.patch
```
I think "basic" tests should follow the default case, i.e. include `Part-of`.

mgorny commented on test/01signoff-missing.sh:
```diff
@@ -43,7 +43,7 @@ cat > trivial.patch <<-EOF
 EOF
 
-! bash "${INITDIR}"/../pram --no-gitconfig -e true -G -I -s ./trivial.patch 2> out.txt
+! bash "${INITDIR}"/../pram --no-gitconfig -e true -G -I -s -P ./trivial.patch 2> out.txt
```
Is `-P` needed here? It shouldn't affect error handling.

mgorny commented on test/11partof.sh:
```diff
@@ -0,0 +1,112 @@
+EOF
+sha256sum -c <<-EOF
+	c95bc3022ee967e117fc7841d9b6597e672d7ef1da7897b2c2692fe8b099911d  data.txt
+EOF
```
Please set up your editor to end files with a `\n`.
